kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kotlinx.collections.immutable.implementations.immutableList
+
+import java.lang.IllegalStateException
+
+internal class PersistentVectorMutableIterator<T>(private val builder: PersistentVectorBuilder<T>,
+                                                  index: Int) : MutableListIterator<T>, AbstractListIterator<T>(index, builder.size) {
+
+    private var expectedModCount = builder.getModCount()
+    private var trieIterator: TrieIterator<T>? = null
+    private var lastIteratedIndex = -1
+
+    init {
+        setupTrieIterator()
+    }
+
+    override fun previous(): T {
+        checkForComodification()
+        checkHasPrevious()
+
+        lastIteratedIndex = index - 1
+
+        val trieIterator = this.trieIterator ?: return builder.tail[--index] as T
+        if (index > trieIterator.size) {
+            return builder.tail[--index - trieIterator.size] as T
+        }
+        index--
+        return trieIterator.previous()
+    }
+
+    override fun next(): T {
+        checkForComodification()
+        checkHasNext()
+
+        lastIteratedIndex = index
+
+        val trieIterator = this.trieIterator ?: return builder.tail[index++] as T
```

Comment:
How comes `tail[index]` doesn't cause index-out-of-bound exception?

Reply:
Ok, I've got that if `trieIterator` is null the list consists of a tail only and index is actually the index in tail. Perhaps worth to document somewhere, as this is unobvious from at first sight.

Reply:
+1